universal-login-web3/lib/index.ts

```diff
@@ -57,26 +36,44 @@ export class ULWeb3Provider implements Provider {
     switch (payload.method) {
       case 'eth_sendTransaction':
         const tx = payload.params[0];
-        this.sdk.execute(tx, WALLET.privateKey)
-          .then(async (ex) => {
-            const mined = await ex.waitToBeMined();
-            callback(null, {
-              id: payload.id,
-              jsonrpc: '2.0',
-              result: mined.transactionHash,
-            });
-          });
+        this.sendTransaction(tx)
+          .then((hash) => callback(null, {
+            id: payload.id,
+            jsonrpc: '2.0',
+            result: hash,
+          }));
         break;
       case 'eth_accounts':
         callback(null, {
           id: payload.id,
           jsonrpc: '2.0',
-          result: [WALLET.contractAddress],
+          result: this.walletService.walletDeployed.get() ? [this.walletService.getDeployedWallet().contractAddress] : [],
         });
         break;
       default:
-        return this.provider.send(payload, callback);
+        return this.provider.send(payload, callback as any);
+    }
+  }
+
+  async sendTransaction(tx: Partial<Message>): Promise<string> {
+    if (!this.walletService.walletDeployed.get()) {
+      this.uiController.requireWallet();
+      await waitFor((x: boolean) => x)(this.walletService.walletDeployed);
+    }
+    return this.executeTransaction(tx);
+  }
+
+  async executeTransaction(tx: Partial<Message>): Promise<string> {
+    const execution = await this.walletService.getDeployedWallet().execute({
+      ...tx,
+      gasLimit: undefined,
+      from: this.walletService.getDeployedWallet().contractAddress,
+    });
+    const mined = await execution.waitToBeMined();
```

Comment:
I think it changes semantics. `sendTransaction` should return immediately after http req was sent.